builder/builder.go

```diff
@@ -852,3 +875,66 @@ func parseUint32(s string) (uint32, error) {
 	}
 	return uint32(parsed), nil
 }
+
+func createCompressedArchive(dst string, srcs ...string) (err error) {
+	defer func() {
+		if err != nil {
+			err = errors.Wrapf(err, "couldn't create compressed archive %s", dst)
+		}
+	}()
+
+	headers := make([]*tar.Header, len(srcs))
+	for i, src := range srcs {
+		var fi os.FileInfo
+		fi, err = os.Lstat(src)
+		if err != nil {
+			return err
+		}
+		if !fi.Mode().IsRegular() {
+			return errors.Errorf("%s has unsupported type of file", src)
+		}
+		headers[i], err = tar.FileInfoHeader(fi, "")
+		if err != nil {
+			return err
+		}
+	}
+
+	archive := &bytes.Buffer{}
+
+	xw, err := swupd.NewExternalWriter(archive, "xz")
+	if err != nil {
+		return err
+	}
+	defer func() {
```

Comment:
I find the error and close handling here a bit forced. Perhaps moving  lines 914 to 933 into its own function and just doing

    xw, err := swupd.NewExternalWriter()
    if err != nil { return err }
    err = some_new_function(xw, headers, srcs)
    cerr := xw.Close()
    if err { return err }
    if cerr { err = cerr /* because of magic err handling! */ ; return err }
    return ioutil.WriteFile()

would make me feel better?
Perhaps move all but 880 to 884 into createCompressedArchiveInternal, and remove the magic handling of "err" as well?

     func createCompressedArchive(dst string, srcs ...string) (err error) {
          err := createCompressedArchiveInternal(dst, srcs...)
          return errors.Wrapf(err, "couldn't create compressed archive %s", dst)
     }

Reply:
Reorganize that function to make it clearer. Split off tar-specific code into a helper function. 

Removed the defer that was wrapping error, and explicit wrap the errors. (I still prefer wrapping using defer, as it is less error prone, but not really that important).